text/1026-net-module.rst

```diff
@@ -0,0 +1,214 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2024-08-14
+    Authors: Scott Trinh <scott@edgedb.com>
+
+=============================
+RFC: EdgeDB Networking Module
+=============================
+
+This RFC outlines the addition of a networking module to EdgeDB’s
+standard library, providing functionality for making HTTP requests and
+sending SMTP messages.
+
+Motivation
+==========
+
+The EdgeDB Auth extension (ext::auth) currently sends emails directly to
+configured SMTP services at certain times in various authentication
+ceremonies and workflows:
+
+-  Sign up with email (sends email verification email)
+-  Password reset (sends password reset email)
+-  Magic link sign up (sends magic link email)
+-  Magic link sign in (sends magic link email)
+
+To allow full customization of emails and provide hooks into
+authentication events generally, we need to support sending webhook
+requests to the application directly. This would enable applications to
+trigger their own logic for sending fully customized and
+internationalized emails, SMS, log analytics, update usage billing
+records, etc.
+
+Specification
+=============
+
+Module Name
+-----------
+
+The new module will be named ``net``.
+
+Utility types
+-------------
+
+Since requests are sent asynchronously, we need a way to track the status of the
+request. This is done by using a ``net::Task`` type to represent a request and
+its current state. Each protocol will have its own ``net::Task`` concrete type.
+
+.. code-block:: edgeql
+
+  scalar type net::TaskState extending std::enums<Pending, InProgress, Complete, Failed>;
+
+  scalar type net::TaskFailure extending std::enums<NetworkError, Timeout>;
+
+  abstract type net::Task {
+    required state: net::TaskState;
+    required created_at: datetime;
+    failure: tuple<failure: net::TaskFailure, message: str>;
+  }
```

Comment:
@1st1 Is this tuple what you had in mind for the failure enum-based error reporting?